compiler/il/OMRNode.cpp

```diff
@@ -2351,6 +2352,56 @@ OMR::Node::isNotCollected()
 
 
 
+/*
+ * This query currently only works on opcodes that can have pinning array pointer,
+ * i.e. aiadd, aiuadd, aladd, aluadd (support for other opcodes can be added if necessary).
+ * This query traverses down the children of the node and try to find if it is originated
+ * from a collectable reference, i.e. anything that is an address to a collected object
+*/
+bool
+OMR::Node::canBeInternalPtrOfObject()
+   {
+   TR::Compilation * comp = TR::comp();
+
+   if (!self()->hasPinningArrayPointer())
+      return false;
+   if (self()->isInternalPointer())
+      return true;
+
+   TR::NodeChecklist visited(comp);
+   TR::Node *curNode = self();
+   // Look for a node with symRef so that we can tell whether self() originates from a collected reference
+   while (curNode)
+      {
+      // Come across a node that has already been checked
+      if (visited.contains(curNode))
+         return false;
+      if (curNode->isInternalPointer())
+         return true;
+      if (curNode->getOpCode().hasSymbolReference())
+         {
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         TR::ILOpCode op = curNode->getOpCode();
+         return op.isNew() ||
+                (op.getDataType() == TR::Address && op.isCall()) ||
+                symbol->isInternalPointer() ||
+                symbol->isCollectedReference();
+         }
+
+      visited.add(curNode);
+      if (curNode->getNumChildren() >= 1)
+         {
+         curNode = curNode->getFirstChild();
+         TR_ASSERT(curNode->getOpCode() != TR::aternary, "canBeInternalPtrOfObject does not work for nodes who originate from a node with data type TR::aternary, node is " POINTER_PRINTF_FORMAT, self());
```

Comment:
I'd prefer this to be a whitelist rather than a black list - if someone adds a new opcode and it isn't accounted for here we should assert I believe. Thoughts?